floors/rule.go

```diff
@@ -0,0 +1,404 @@
+	var value string
+	if request.Site != nil {
+		value = request.Site.Domain
+	} else {
+		value = request.App.Domain
+	}
+	return value
+}
+
+func getPublisherDomain(request *openrtb2.BidRequest) string {
+	var value string
+	if request.Site != nil {
+		value = request.Site.Publisher.Domain
+	} else {
+		value = request.App.Publisher.Domain
+	}
+	return value
+}
+
+func getBundle(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.App != nil {
+		value = request.App.Bundle
+	}
+	return value
+}
+
+func getgptslot(imp openrtb2.Imp) string {
+	var value string
+	adsname, err := jsonparser.GetString(imp.Ext, "data", "adserver", "name")
+	if err == nil && adsname == "gam" {
+		gptSlot, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "adslot")
+		if gptSlot != "" {
+			value = gptSlot
+		}
+	} else {
+		value = getpbadslot(imp)
+	}
+	return value
+}
+
+func extractChanelNameFromBidRequestExt(bidRequest *openrtb2.BidRequest) string {
+	requestExt := &openrtb_ext.ExtRequest{}
+	if bidRequest == nil {
+		return CATCH_ALL
+	}
+
+	if len(bidRequest.Ext) > 0 {
+		err := json.Unmarshal(bidRequest.Ext, &requestExt)
+		if err != nil {
+			return CATCH_ALL
+		}
+	}
+
+	if requestExt.Prebid.Channel != nil {
+		return requestExt.Prebid.Channel.Name
+	}
+	return CATCH_ALL
+}
+
+func getpbadslot(imp openrtb2.Imp) string {
+	value := CATCH_ALL
+	pbAdSlot, err := jsonparser.GetString(imp.Ext, "data", "pbadslot")
+	if err == nil {
+		value = pbAdSlot
+	}
+	return value
+}
+
+func getAdUnitCode(imp openrtb2.Imp) string {
+	adUnitCode := CATCH_ALL
+	gpId, err := jsonparser.GetString(imp.Ext, "gpid")
+	if err == nil && gpId != "" {
+		return gpId
+	}
+
+	if imp.TagID != "" {
+		return imp.TagID
+	}
+
+	pbAdSlot, err := jsonparser.GetString(imp.Ext, "data", "pbadslot")
+	if err == nil && pbAdSlot != "" {
+		return pbAdSlot
+	}
+
+	storedrequestID, err := jsonparser.GetString(imp.Ext, "prebid", "storedrequest", "id")
+	if err == nil && storedrequestID != "" {
+		return storedrequestID
+	}
+	return adUnitCode
+}
+
+func isMobileDevice(userAgent string) bool {
+	isMobile, err := regexp.MatchString("(?i)Phone|iPhone|Android|Mobile", userAgent)
+	if err != nil {
+		return false
+	}
+	return isMobile
+}
+
+func isTabletDevice(userAgent string) bool {
+	isTablet, err := regexp.MatchString("(?i)tablet|iPad|Windows NT", userAgent)
+	if err != nil {
+		return false
+	}
+	return isTablet
+}
+
+func prepareRuleCombinations(keys []string, numSchemaFields int, delimiter string) []string {
+	var subset []string
+	var comb []int
+	var desiredkeys [][]string
+	var ruleKeys []string
+
+	segNum := 1 << numSchemaFields
+	for i := 0; i < numSchemaFields; i++ {
+		subset = append(subset, strings.ToLower(keys[i]))
+		comb = append(comb, i)
+	}
+	desiredkeys = append(desiredkeys, subset)
+	for numWildCart := 1; numWildCart <= numSchemaFields; numWildCart++ {
+		newComb := generateCombinations(comb, numWildCart, segNum)
+		for i := 0; i < len(newComb); i++ {
```

Comment:
Time complexity is O(2^N) complexity where N is the number of schema fields, in actual use case N could be less than 12.
It would be great if we get ideas on  more efficient algorithm.
We can prepare combination set for each N  in first request and put it in cache and in 2nd request onwards we can use cached combination, however rule keys combination should be done to each request.